lightweight-term-count-update.php

```diff
@@ -12,79 +12,219 @@
+	 */
+	public static function instance() {
+		if ( ! isset( self::$instance ) ) {
+			self::$instance = new static;
+			self::$instance->setup();
+		}
+		return self::$instance;
 	}
 
-	public function init() {
+	/**
+	 * Setup the singleton.
+	 */
+	public function setup() {
+		// Prevent core from counting terms.
+		wp_defer_term_counting( true );
 		remove_action( 'transition_post_status', '_update_term_count_on_transition_post_status' );
-		add_action( 'transition_post_status', array( $this, 'quick_update_terms_count' ), 10, 3 );
-		$this->counted_stati = apply_filters( 'LTCU_counted_stati',  $this->counted_stati );
+
+		$this->counted_statuses = apply_filters( 'ltcu_counted_statuses', $this->counted_statuses );
+		add_action( 'transition_post_status', array( $this, 'transition_post_status' ), 10, 3 );
+		add_action( 'added_term_relationship', array( $this, 'added_term_relationship' ), 10, 3 );
+		add_action( 'deleted_term_relationships', array( $this, 'deleted_term_relationships' ), 10, 3 );
 	}
 
-	public function quick_update_terms_count( $new, $old, $post ) {
-		global $wpdb;
+	/**
+	 * When a term relationship is added, increment the term count.
+	 *
+	 * @param int    $object_id Object ID.
+	 * @param int    $tt_id     Single term taxonomy ID.
+	 * @param string $taxonomy  Taxonomy slug.
+	 */
+	public function added_term_relationship( $object_id, $tt_id, $taxonomy ) {
+		$post = get_post( $object_id );
+		if ( in_array( $post->post_status, $this->counted_statuses, true ) ) {
+			$this->quick_update_terms_count( $post, (array) $tt_id, $taxonomy, 'increment' );
+		}
+	}
+
+	/**
+	 * When a term relationship is deleted, decrement the term count.
+	 *
+	 * @param int    $object_id Object ID.
+	 * @param array  $tt_ids    Array of term taxonomy IDs.
+	 * @param string $taxonomy  Taxonomy slug.
+	 */
+	public function deleted_term_relationships( $object_id, $tt_ids, $taxonomy ) {
+		$post = get_post( $object_id );
+		$this->quick_update_terms_count( $post, $tt_ids, $taxonomy, 'decrement' );
+	}
+
+	/**
+	 * When a post transitions, increment or decrement term counts as
+	 * appropriate.
+	 *
+	 * @param  string   $new_status New post status.
+	 * @param  string   $old_status Old post status.
+	 * @param  \WP_Post $post       Post being transitioned.
+	 */
+	public function transition_post_status( $new_status, $old_status, $post ) {
+		foreach ( (array) get_object_taxonomies( $post->post_type ) as $taxonomy ) {
+			$tt_ids = wp_get_object_terms( $post->ID, $taxonomy, array(
+				'fields' => 'tt_ids',
+			) );
+
+			if ( ! empty( $tt_ids ) ) {
+				$this->quick_update_terms_count(
+					$post,
+					$tt_ids,
+					$taxonomy,
+					$this->transition_type( $new_status, $old_status )
+				);
+			}
+		}
+
+		// For non-attachments, let's check if there are any attachment children
+		// with inherited post status -- if so those will need to be re-counted.
+		if ( 'attachment' !== $post->post_type ) {
+			$attachments = new WP_Query( array(
```

Comment:
> This returns only latest 10 attachments. I'm not sure whether setting the `posts_per_page` should be set to `-1` or whether we should come up with some clever direct SQL. Thoughts?

I was wondering that too. I'm not sure what to do here, to be honest, but I do have some suggestions. To maintain parity with core, we would need to set `'posts_per_page' => -1` to catch all possible attachments, but that could theoretically be worse than `_update_post_term_count()`. Here are a few ideas:

1. We could query just for `'fields' => 'ids'`. All we ever use of the post objects are the `ID` and `post_type`, and we already know the post type is `attachment`. This would keep the query light, but it still means that every attachment could theoretically trigger another query to update the term count, so if a post has 100k attachments we could theoretically be looking at 100,002 total sql queries.
     * A sub-option here would be to refactor the code to be able to pool the attachment queries into one query. That would turn the 100,002 queries down to 1 or 2 queries, where one of them would have 100k+ `tt_id`s in the `IN()` clause.
2. We could do a quick count synchronously for the post and then setup an asynchronous cron task to count attachments, if there are any (or if there are more than, say, 100).
3. Count the attachments, and if the number is high, just let `_update_post_term_count()` run as it usually would. Additionally, we could add a filter to optionally override it, and then people can intervene with custom code if it's a real problem for them, and do something else like (1) or (2) or offload the query to Elasticsearch or something.

(1) or (1a) seem like the best options to me of what I've considered. (2) feels over-engineered for the 0.001%, or whatever hilariously low percentage it would be, of use cases that could be problematic. (3) is a fine option too -- we fixed all the use cases most people will ever encounter, and for that one in a million, we simply left it as-is but opened a door for that site to easily solve their problem.